core/include/vecmem/memory/details/arena_memory_resource.hpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021-2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Local include(s).
+#include "vecmem/memory/details/memory_resource_base.hpp"
+#include "vecmem/memory/memory_resource.hpp"
+#include "vecmem/vecmem_core_export.hpp"
+
+namespace vecmem::details {
+
+// Forward declaration(s).
+class arena_memory_resource_impl;
+
+/// Memory resource implementing an arena allocation scheme
+class VECMEM_CORE_EXPORT arena_memory_resource : public memory_resource_base {
+
+public:
+    /// Construct the memory resource on top of an upstream memory resource
+    ///
+    /// @param[in] upstream The @c vecmem::memory_resource to use for "upstream"
+    ///                     memory allocations
+    /// @param[in] initial_size Initial memory memory allocation from
+    ///                         @c upstream
+    /// @param[in] maximum_size The maximal allowed allocation from @c upstream
+    ///
+    arena_memory_resource(memory_resource& upstream, std::size_t initial_size,
+                          std::size_t maximum_size);
+    /// Move constructor
+    arena_memory_resource(arena_memory_resource&& parent);
+    /// Disallow copying the memory resource
+    arena_memory_resource(const arena_memory_resource&) = delete;
+
+    /// Destructor
+    ~arena_memory_resource();
+
+    /// Move assignment operator
+    arena_memory_resource& operator=(arena_memory_resource&& rhs);
+    /// Disallow copying the memory resource
+    arena_memory_resource& operator=(const arena_memory_resource&) = delete;
+
+protected:
+    /// @name Function(s) implementing @c vecmem::details::memory_resource_base
+    /// @{
+
+    /// Allocate memory in the arena
+    virtual void* mr_allocate(std::size_t bytes, std::size_t) override;
+    /// De-allocate a previously allocated memory block
+    virtual void mr_deallocate(void* p, std::size_t bytes,
+                               std::size_t) override;
+
+    /// @}
+
+private:
+    /// Object performing the heavy lifting for the memory resource
+    arena_memory_resource_impl* m_impl;
```

Comment:
You mentioned this in the description already, but it's really unfortunately not being able to use smart-pointers for the PIMPL here. If there's absolutely now way to get MSVC to swallow that, fine, but it would really be a shame.

Reply:
I *think* MSVC should actually allow you to export `std::unique_ptr` in principle.

Reply:
It's a pretty complicated class. I looked at it again for a bit yesterday in `libstdc++`. And I can sort of see Microsoft's point after that.

Since what they want is that after you build your DLL with MSVC XX, that DLL would be usable as-is with MSVC YY. And that is not guaranteed with `std::unique_ptr` in the "memory footprint" of your classes. 🤔 Note that various of our memory resources still receive references to STL types. I guess one could argue there that as long as these are references, the actual implementation in the client that provides the reference can be different from the implementation that the VecMem code saw when it was compiled. As long as the API didn't change, this would still keep the code working.

However the logic of even that breaks down with the `vecmem::instrumenting_memory_resource::get_events()` function. 😦 Since that returns an `std::vector` of objects. And in case `std::vector` changes its internal representation between the time that `vecmem_core.dll` is built, and that the user makes use of it, that function would change its ABI. 🤔 

But, MSVC doesn't complain about it, so I'm not going to worry about it for now... 😛

Reply:
But does it actually **fail** due to `unique_ptr` then? I see the argument about broken ABI compat because it's not guaranteed, but that's not something that we want to try to even guarantee to that degree.

If we can make MSVC shut up about `unique_ptr`, I'd be in favor of technically not having real ABI stability, but using the smart pointer instead of manual deletes.

Reply:
It would reduce the number of warnings being suppressed but I see your point. 

Another option would be to not export the class altogether, but only the methods that we actually want to support. In that case I think MSVC would not complain about `unique_ptr`. That's not super great either, of course.

Ok, then I guess we should follow this up with a custom unique smart pointer to hide all the `delete`s and raw `new` calls, in my opinion.

Reply:
As long as the client needs to be able to construct the memory resource object, they need to know how to construct an `std::unique_ptr`. (Or at least how much memory to reserve on the stack for making that happen.) We can't suppress exporting this aspect of the class. 😦

Note that since it's a private member, technically it's only the vecmem code that actually interacts with the `std::unique_ptr` ABI. I think it's more of a technical thing of how the Windows ABI communicates how much memory is needed for a given object, through whatever mechanism they have for exporting symbols. 🤔

**Very** long story short, let's add a simple smart pointer for this purpose in a follow-up PR.